### PR TITLE
OsCore-driver-runner split

### PR DESCRIPTION
**Pull request by michalusio**
Moved the core and drivers to their separate directories.
The drivers are registered by the runner to the core, using an extern function that returns Driver information.
The logger is a mutable global reference but we could do something about it in the next PRs (as we probably want some kind of driver<->kernel communication anyway).

Needs to have the tests set up properly but I don't know how to do that.

### CODE REVIEW

File: os_core/src/structures/driver.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+use super::kernel_information::KernelInformation;
+
+pub type Registrator = extern "C" fn(KernelInformation) -> Driver;
+
+#[derive(Clone, Copy)]
+#[repr(C)]
+pub struct Driver {
+    pub serial: [u8; 16],
```
Comment (0xffset) [inline os_core/src/structures/driver.rs:8]: also where does the serial array come from that you defined in all the devices? Did you just make it up?

**Review by 0xffset — CHANGES_REQUESTED** [review]

File: os_core/src/structures/driver.rs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+use super::kernel_information::KernelInformation;
+
+pub type Registrator = extern "C" fn(KernelInformation) -> Driver;
```
[the remark this answers is not in the document]
Comment (michalusio) [reply os_core/src/structures/driver.rs:3]: In the future we will want to have the drivers be loaded from the file system, so they have to have the C function signature.
It's there so we don't forget that.

File: os_core/src/os_core.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    };
+    let mut frame_allocator = unsafe { BootInfoFrameAllocator::init(&boot_info.memory_regions) };
+
+    memory::heap::init_heap(&mut mapper, &mut frame_allocator).expect("heap initialization failed");
+
+    kernel_info
+}
+
+pub fn reload_drivers(kernel_info: KernelInformation) {
+    let mut initialized_drivers = INITIALIZED_DRIVERS.lock();
```
[the remark this answers is not in the document]
Comment (michalusio) [reply os_core/src/os_core.rs:64]: I can make it a Vec now that we have heap allocation.

File: os_core/src/structures/driver.rs (hunk 1 above)
Comment (michalusio) [reply os_core/src/structures/driver.rs:8]: I just made it up. We want the drivers to have some kind of signature id (should have named the property like that tbh).

File: os_core/src/structures/driver.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (0xffset) [reply os_core/src/structures/driver.rs:3]: ok sounds about right. Can you add a comment something along the lines of `extern C for future compatibility`

File: os_core/src/structures/driver.rs (hunk 1 above)
Comment (0xffset) [reply os_core/src/structures/driver.rs:8]: Okay, figured so much just wanted to make sure

File: os_core/src/os_core.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (0xffset) [reply os_core/src/os_core.rs:64]: Okay, i thought maybe it had a reason (like hardware limits of accessible devices or something)

**Review by 0xffset — APPROVED** [review]

**Pull request merged**